ctapipe/reco/hillas_intersection.py

```diff
@@ -495,25 +497,19 @@ def reconstruct_xmax(
             np.array(ty),
         )
         weight = np.array(amp)
-        mean_height = np.sum(height * weight) / np.sum(weight)
+        mean_distance = np.sum(height * weight) / np.sum(weight)
 
         # This value is height above telescope in the tilted system,
         # we should convert to height above ground
-        mean_height *= np.cos(zen)
+        mean_height = mean_distance * np.cos(zen.to_value(u.rad))
 
         # Add on the height of the detector above sea level
-        mean_height += 2100  # TODO: replace with instrument info
+        mean_height += self.subarray.reference_location.geodetic.height.to_value(u.m)
 
         if mean_height > 100000 or np.isnan(mean_height):
```

Comment:
Checking for nan doesn't make sense anymore if you set to nan.

Reply:
Replace magic number with something like "H_MAX_UPPER_LIMIT = 100000"?

Reply:
done